### PR TITLE
Fix the difference between aws_waf_byte_match_set and aws_wafregional_byte_match_set

### PR DESCRIPTION
**Pull request by chroju**
Changes proposed in this pull request:

Some aws waf and aws waf-regional API have same attributes each other. However, Terraform `aws_waf_***` resources and `aws_wafregional_***` resources have same attributes with the different name, one is singular and the other is plural.  Although this is mentioned a little in #4137 , this discrepancy can also be seen in several other resources about AWS WAF(Regional).

This PR will fix such inconsistencies. I think it would be better to unify in singular names, because these attributes type is not `list` but `map` .

Output from acceptance testing:

WIP

```
$ make testacc TEST=./aws TESTARGS='-run=TestAccAWSWafByteMatchSet*'
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./aws -v -run=TestAccAWSWafByteMatchSet* -timeout 120m
=== RUN   TestAccAWSWafByteMatchSet_basic
--- PASS: TestAccAWSWafByteMatchSet_basic (42.07s)
=== RUN   TestAccAWSWafByteMatchSet_changeNameForceNew
--- PASS: TestAccAWSWafByteMatchSet_changeNameForceNew (77.94s)
=== RUN   TestAccAWSWafByteMatchSet_changeTuple
--- PASS: TestAccAWSWafByteMatchSet_changeTuple (72.65s)
=== RUN   TestAccAWSWafByteMatchSet_noTuples
--- PASS: TestAccAWSWafByteMatchSet_noTuples (37.13s)
=== RUN   TestAccAWSWafByteMatchSet_disappears
--- PASS: TestAccAWSWafByteMatchSet_disappears (35.07s)
PASS
ok      github.com/chroju/terraform-provider-aws/aws    264.901s
```

### CODE REVIEW

File: aws/resource_aws_waf_byte_match_set.go [hunk 1]
```diff
@@ -24,7 +24,7 @@ func resourceAwsWafByteMatchSet() *schema.Resource {
 				Required: true,
 				ForceNew: true,
 			},
-			"byte_match_tuples": &schema.Schema{
```
Comment (bflad) [inline aws/resource_aws_waf_byte_match_set.go]: We need to duplicate the entire existing `byte_match_tuples` attribute into the new `byte_match_tuple` attribute. Then in the existing `byte_match_tupes` attribute, add `ConflictsWith` and `Deprecation` fields, e.g.

```go
ConflictsWith: []string{"byte_match_tuple"},
Deprecation: "use `byte_match_tuple` instead",
```

File: aws/resource_aws_waf_byte_match_set.go [hunk 2]
```diff
@@ -107,7 +107,7 @@ func resourceAwsWafByteMatchSetRead(d *schema.ResourceData, meta interface{}) er
 	}
 
 	d.Set("name", resp.ByteMatchSet.Name)
-	d.Set("byte_match_tuples", flattenWafByteMatchTuples(resp.ByteMatchSet.ByteMatchTuples))
```
Comment (bflad) [inline aws/resource_aws_waf_byte_match_set.go]: When renaming attributes, we need to ensure both the old and new are still read into the Terraform state.

File: aws/resource_aws_waf_byte_match_set.go [hunk 3]
```diff
@@ -117,8 +117,8 @@ func resourceAwsWafByteMatchSetUpdate(d *schema.ResourceData, meta interface{})
 
 	log.Printf("[INFO] Updating ByteMatchSet: %s", d.Get("name").(string))
 
-	if d.HasChange("byte_match_tuples") {
```
Comment (bflad) [inline aws/resource_aws_waf_byte_match_set.go]: The existing logic for `byte_match_tuples` must be preserved in addition to the new `byte_match_tuple`

File: aws/resource_aws_waf_byte_match_set.go [hunk 4]
```diff
@@ -24,7 +24,7 @@ func resourceAwsWafByteMatchSet() *schema.Resource {
 				Required: true,
 				ForceNew: true,
 			},
-			"byte_match_tuples": &schema.Schema{
+			"byte_match_tuple": &schema.Schema{
```
Comment (bflad) [inline aws/resource_aws_waf_byte_match_set.go]: The documentation in `website/docs/r/waf_byte_match_set.html.markdown` should add the new attribute and add a **deprecated** notice to the existing attribute.

File: aws/resource_aws_waf_rule_test.go [hunk 5]
```diff
@@ -402,15 +402,15 @@ func testAccAWSWafRuleConfig_changePredicates(name string) string {
 	return fmt.Sprintf(`
 resource "aws_waf_ipset" "ipset" {
   name = "%s"
-  ip_set_descriptors {
```
Comment (bflad) [inline aws/resource_aws_waf_rule_test.go]: Was this change intentional?

**Comment by bflad** [comment]
For what its worth, there can be multiple ByteMatchSet tuples as the attribute is `schema.TypeSet` and the `updateByteMatchSetResource` function works with multiple tuples. It seems this attribute should be plural.

**Comment by chroju** [comment]
@bflad Thanks for your review and comments.

I have noticed that this is so complex and breaking change than I thought. There are much more inconsistent attributes about AWS WAF and AWS WAFRegional, and it is debatable whether each of them should be plural or singular.

So, at first I would fix this PR not about these all inconsistent attributes but only `byte_match_tuples` . Also I will make the list of inconsistent attributes at #4137 once. Is this OK ?

File: aws/resource_aws_waf_rule_test.go (hunk 5 above)
Comment (chroju) [reply aws/resource_aws_waf_rule_test.go]: Sorry, this is my mistake. Too mach changes were in the commit.

**Comment by bflad** [comment]
This has been released in [version 1.27.0 of the AWS provider](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#1270-july-11-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.

**Comment by nbrys** [comment]
After changing 

`byte_match_tuple {
    text_transformation   = "NONE"
    target_string         = "/favicon.png"
    positional_constraint = "EXACTLY"

    field_to_match {
      type = "URI"
    }
  }`

into:
`byte_match_tuples {
    text_transformation   = "NONE"
    target_string         = "/favicon.png"
    positional_constraint = "EXACTLY"

    field_to_match {
      type = "URI"
    }
  }`

I get the following change in terraform:
  byte_match_tuple.511953057.field_to_match.#:                 "1" => "0"
  byte_match_tuple.511953057.field_to_match.3756326843.data:   "" => ""
  byte_match_tuple.511953057.field_to_match.3756326843.type:   "URI" => ""
  byte_match_tuple.511953057.positional_constraint:            "EXACTLY" => ""
  byte_match_tuple.511953057.target_string:                    "/favicon.ico" => ""
  byte_match_tuple.511953057.text_transformation:              "NONE" => ""
  byte_match_tuples.#:                                         "0" => "1"
  byte_match_tuples.511953057.field_to_match.#:                "0" => "1"
  byte_match_tuples.511953057.field_to_match.3756326843.data:  "" => ""
  byte_match_tuples.511953057.field_to_match.3756326843.type:  "" => "URI"
  byte_match_tuples.511953057.positional_constraint:           "" => "EXACTLY"
  byte_match_tuples.511953057.target_string:                   "" => "/favicon.ico"
  byte_match_tuples.511953057.text_transformation:             "" => "NONE"

But I end up with an empty filter in WAF
